### PR TITLE
Remove Geth Bindings From Prysm

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Execution Code Cleanup

**What does this PR do? Why is it needed?**

Post merge, prysm retrieves raw headers and blocks from geth without using go bindings from the go-ethereum repo. This PR extends this change across all our execution interfacing code so that we can remove some unnecessary interfaces along with supporting gnosis chain changes.

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/execution/engine_client.go [hunk 1]
```diff
@@ -354,6 +355,24 @@ func (s *Service) ExecutionBlocksByHashes(ctx context.Context, hashes []common.H
 	return execBlks, nil
 }
 
+func (s *Service) HeaderByHash(ctx context.Context, hash common.Hash) (*pb.ExecutionBlock, error) {
```
Comment (rkapka) [inline beacon-chain/execution/engine_client.go:358]: Missing doc

File: beacon-chain/execution/engine_client.go [hunk 2]
```diff
@@ -354,6 +355,24 @@ func (s *Service) ExecutionBlocksByHashes(ctx context.Context, hashes []common.H
...
+	var blk *pb.ExecutionBlock
+	err := s.rpcClient.CallContext(ctx, &blk, "eth_getBlockByHash", hash, false /* no transactions */)
+	if err == nil && blk == nil {
+		err = ethereum.NotFound
+	}
+	return blk, err
+}
+
+func (s *Service) HeaderByNumber(ctx context.Context, number *big.Int) (*pb.ExecutionBlock, error) {
```
Comment (rkapka) [inline beacon-chain/execution/engine_client.go:367]: Missing doc

File: beacon-chain/execution/testing/mock_execution_chain.go [hunk 3]
```diff
@@ -140,12 +142,89 @@ func (m *Chain) ETH1ConnectionErrors() []error {
...
 type RPCClient struct {
-	Backend *backends.SimulatedBackend
+	Backend     *backends.SimulatedBackend
+	BlockNumMap map[uint64]*pb.ExecutionBlock
 }
 
 func (*RPCClient) Close() {}
 
-func (*RPCClient) CallContext(_ context.Context, _ interface{}, _ string, _ ...interface{}) error {
+func (r *RPCClient) CallContext(ctx context.Context, obj interface{}, methodName string, args ...interface{}) error {
```
Comment (rkapka) [inline beacon-chain/execution/testing/mock_execution_chain.go:151]: Missing doc

**Comment by prestonvanloon** [comment]
I don't understand the rationale here. In any case, can we simplify this further?  
Keeping a full block in the cache or passing that around seems like extra mem usage that is unnecessary. 

We discussed offline that we only need the hash, timestamp and block number. Some subset of the block.

Similar to how LH does it https://github.com/sigp/lighthouse/blob/6d5a2b509fac7b6ffe693866f58ba49989f946d7/common/eth2/src/lighthouse.rs#L314

They have a subset of the block struct they are using. I imagine this is far less memory usage than a full eth1 block with everything in it.

**Comment by prestonvanloon** [comment]
OK -- actually I remember why. The go-ethereum structs recompute the hash when you call `Hash()` and this is incompatible with gnosis structs as they are different. In our case, we can still simplify further by only managing a subset of the response object.

File: beacon-chain/execution/testing/mock_execution_chain.go (hunk 3 above)
Comment (nisdas) [reply beacon-chain/execution/testing/mock_execution_chain.go:150]: This is a test method

File: beacon-chain/execution/engine_client.go [hunk 4]
```diff
@@ -354,6 +356,26 @@ func (s *Service) ExecutionBlocksByHashes(ctx context.Context, hashes []common.H
...
+// HeaderByNumber returns the relevant header details for the provided block number.
+func (s *Service) HeaderByNumber(ctx context.Context, number *big.Int) (*types.HeaderInfo, error) {
+	var hdr *types.HeaderInfo
+	err := s.rpcClient.CallContext(ctx, &hdr, ExecutionBlockByNumberMethod, toBlockNumArg(number), false /* no transactions */)
+	if err == nil && hdr == nil {
+		err = ethereum.NotFound
+	}
+	return hdr, err
+}
+
```
Comment (prestonvanloon) [inline beacon-chain/execution/engine_client.go:378]: Do we have any way to test these? Mock rpc client or something? The functionality for returning `ethereum.NotFound` is something worth testing, unless it is very difficult to do so.

File: beacon-chain/execution/block_reader.go [hunk 5]
```diff
@@ -207,20 +207,17 @@ func (s *Service) retrieveHeaderInfo(ctx context.Context, bNum uint64) (*types.H
...
 		}
-		if blk == nil {
+		if hdr == nil {
 			return nil, errors.Errorf("header with the number %d does not exist", bNum)
 		}
-		if err := s.headerCache.AddHeader(blk); err != nil {
-			return nil, err
-		}
-		info, err = types.HeaderToHeaderInfo(blk)
-		if err != nil {
```
Comment (prestonvanloon) [inline beacon-chain/execution/block_reader.go:221]: Oh cool, we were already storing a subset of the block information in the caches.

File: beacon-chain/execution/service.go [hunk 6]
```diff
@@ -101,8 +100,6 @@ type Chain interface {
 // fetching eth1 data from the clients.
 type RPCDataFetcher interface {
 	Close()
-	HeaderByNumber(ctx context.Context, number *big.Int) (*gethTypes.Header, error)
-	HeaderByHash(ctx context.Context, hash common.Hash) (*gethTypes.Header, error)
 }
```
Comment (prestonvanloon) [inline beacon-chain/execution/service.go:103]: Delete this whole thing?

File: beacon-chain/execution/service_test.go [hunk 7]
```diff
@@ -82,7 +84,7 @@ func (g *goodNotifier) StateFeed() *event.Feed {
 
 type goodFetcher struct {
 	backend     *backends.SimulatedBackend
-	blockNumMap map[uint64]*gethTypes.Header
+	blockNumMap map[uint64]*pb.ExecutionBlock
```
Comment (prestonvanloon) [inline beacon-chain/execution/service_test.go:87]: This one is different than the mock rpcClient?

```go
// beacon-chain/execution/testing/mock_execution_chain.go
// RPCClient defines the mock rpc client.
type RPCClient struct {
	Backend *backends.SimulatedBackend
	Backend     *backends.SimulatedBackend
	BlockNumMap map[uint64]*types.HeaderInfo // not pb.ExecutionBlock
}
```

Also, can you just use that mock RPCClient?

Comment (nisdas) [reply beacon-chain/execution/service_test.go:87]: We don't need the whole mock actually, just deleting it now

File: beacon-chain/execution/service.go [hunk 8]
```diff
@@ -124,6 +116,18 @@ func (RPCClientEmpty) CallContext(context.Context, interface{}, string, ...inter
...
+
+func (RPCClientBad) Close() {}
+func (RPCClientBad) BatchCall([]gethRPC.BatchElem) error {
+	return errors.New("rpc client is not initialized")
+}
+
+func (RPCClientBad) CallContext(context.Context, interface{}, string, ...interface{}) error {
+	return ethereum.NotFound
+}
+
```
Comment (prestonvanloon) [inline beacon-chain/execution/service.go:130]: This looks like test code?

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**